pandas/tests/indexes/multi/test_constructor.py

```diff
@@ -473,6 +473,44 @@ def test_from_product_datetimeindex():
     tm.assert_numpy_array_equal(mi.values, etalon)
 
 
+def test_from_product_infer_names():
```

Comment:
You can combine this all into a single parametrized test:
```python
@pytest.mark.parametrize("a, b, expected_name", [
    (pd.Series([1, 2, 3], name="foo"), pd.Series(["a", "b"], name="bar"), ["foo", "bar"]),
    (pd.Series([1, 2, 3], name="foo"), ["a", "b"], ["foo", None]),
    ([1, 2, 3], ["a", "b"], None)])
def test_from_product_infers_partial_names(a, b, expected_names):
    result = MultiIndex.from_product([a, b])
    expected = MultiIndex(
        levels=[[1, 2, 3], ["a", "b"]],
        codes=[[0, 0, 1, 1, 2, 2], [0, 1, 0, 1, 0, 1]],
        names=expected_names,
    )
    tm.assert_index_equal(result, expected)
```

Reply:
Really cool feature, need to look more into this, but the test has been compacted.